### PR TITLE
MessageHub health check

### PR DESCRIPTION
**Pull request by michaelmajgaard**
#469

### CODE REVIEW

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            var cosmosConnectionString = _cosmosDatabaseConfig.ConnectionsString;
+            var serviceBusConnectionString = _serviceBusConfig.DataAvailableQueueConnectionString;
+
+            var results = await _health.NewFluentHealthVerifier
+                .VerifyCosmosDatabase("MESSAGEHUB_COSMOS_DB", cosmosConnectionString, _cosmosDatabaseConfig.MessageHubDatabaseId)
+                .VerifyCosmosDatabase("MESSAGEHUB_COSMOS_LOG_DB", cosmosConnectionString, _cosmosDatabaseConfig.LogDatabaseId)
+                .VerifyMessageBus("MESSAGE_HUB_DATAAVAILABLE_QUEUE", serviceBusConnectionString, _serviceBusConfig.DataAvailableQueueName)
+                .VerifyMessageBus("MESSAGE_HUB_DATAAVAILABLE_ARCHIVE_QUEUE", serviceBusConnectionString, _serviceBusConfig.DequeueCleanUpQueueName)
+                .VerifySqlDatabase("ACTOR_DB", _actorDbConfig.ConnectionString)
+                .RunInParallelAsync().ConfigureAwait(false);
```
Comment (defectiveAi) [inline source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs:61]: Verify is an action, but you are only adding stuff. Suggested names:
```
_health
.Verification
.AddCosmosDatabase...
.AddServiceBusQueue
.AddSqlDatabase
.VerifyInParallelAsync
```

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        private static HttpStatusCode CalculateStatusCode(IDictionary<string, bool> healthChecks)
+        {
+            return healthChecks.All(x => x.Value) ? HttpStatusCode.OK : HttpStatusCode.ServiceUnavailable;
+        }
+
+        private static string Print(IDictionary<string, bool> healthChecks)
+        {
+            const string OKMessage = "OK";
```
Comment (defectiveAi) [inline source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs:77]: I expect naming squigglies here. okMessage? UnavailableMessage?

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs (hunk 1 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs:62]: makes sense, fixed

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            const string unavailableMessage = "unavailable";
+
+            var len = healthChecks.Any() ? healthChecks.Max(x => x.Key.Length + (x.Value ? OKMessage : unavailableMessage).Length + 2) : 0;
+            var grouped = healthChecks.OrderBy(x => x.Key).GroupBy(x => x.Value).ToDictionary(x => x.Key, x => string.Join("\n", x.Select(x => $"{x.Key}: {(x.Value ? OKMessage : unavailableMessage)}")));
+            var headerBar = new string('=', len);
+            return
+                $"{(grouped.TryGetValue(false, out var f) ? $"Failed checks:\n{headerBar}\n{f}\n\n" : string.Empty)}" +
+                $"{(grouped.TryGetValue(true, out var t) ? $"Succeeded checks:\n{headerBar}\n{t}" : string.Empty)}";
```
Comment (defectiveAi) [inline source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs:86]: Hard to read, not sure what is happening...

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/CosmosDatabaseVerifier.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace Energinet.DataHub.PostOffice.EntryPoint.Operations.HealthCheck
+{
+    public sealed class CosmosDatabaseVerifier : ICosmosDatabaseVerifier
+    {
+        public async Task<bool> VerifyAsync(string connectionString, string name)
+        {
+            try
+            {
+                using var cosmosClient = new CosmosClient(connectionString);
```
Comment (defectiveAi) [inline source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/CosmosDatabaseVerifier.cs:27]: Is there an await using?

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs (hunk 2 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs:78]: Both 'OkMessage' and 'OKMessage' is allowed, but not 'okMessage'. I'll change it to 'OkMessage'

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/FluentHealth.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            return this;
+        }
+
+        public IFluentHealth VerifyMessageBus(string verficationKey, string connectionString, string name)
+        {
+            Verify(verficationKey, () => _serviceBusVerifier.VerifyAsync(connectionString, name));
+            return this;
+        }
+
+        public async Task<IDictionary<string, bool>> RunAsync()
```
Comment (defectiveAi) [inline source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/FluentHealth.cs:55]: The mapped result should probably be an IEnumerable<> of some obj/tuple (string Key, bool VerifiedSuccessfully). Not sure why it is a dictionary.

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs (hunk 3 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Functions/HealthFunction.cs:86]: It makes pretty text formatting. It's solid :)

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/CosmosDatabaseVerifier.cs (hunk 4 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/CosmosDatabaseVerifier.cs:27]: Nope

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/FluentHealth.cs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            var results = new ConcurrentDictionary<string, (int Order, bool Result)>();
+
+            await Task.WhenAll(_verifications.Select(x => RunVerifierAsync(results, x.Key, x.Value))).ConfigureAwait(false);
+
+            return MapResults(results);
+        }
+
+        private static Dictionary<string, bool> MapResults(IDictionary<string, (int Order, bool Result)> results)
+        {
+            return results.OrderBy(x => x.Value.Order).ToDictionary(x => x.Key, x => x.Value.Result);
```
Comment (defectiveAi) [inline source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/FluentHealth.cs:78]: What is the point of order, if you use a dictionary, which will internally sort/enumerate by key anyway?

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/Health.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+        private readonly IServiceBusQueueVerifier _serviceBusVerifier;
+
+        public Health(ICosmosDatabaseVerifier cosmosDatabaseVerifier, ISqlDatabaseVerifier sqlDatabaseVerifier, IServiceBusQueueVerifier serviceBusVerifier)
+        {
+            _cosmosDatabaseVerifier = cosmosDatabaseVerifier;
+            _sqlDatabaseVerifier = sqlDatabaseVerifier;
+            _serviceBusVerifier = serviceBusVerifier;
+        }
+
+        public IFluentHealth NewFluentHealthVerifier => new FluentHealth(_cosmosDatabaseVerifier, _sqlDatabaseVerifier, _serviceBusVerifier);
```
Comment (defectiveAi) [inline source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/Health.cs]: Looks like a method to me. Something like CreateVerifier(). A property with a verb in name is sus.

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Program.cs [hunk 8]
```diff
@@ -33,6 +34,8 @@ await using (startup.ConfigureAwait(false))
                     .ConfigureFunctionsWorkerDefaults(options =>
                     {
                         options.UseMiddleware<SimpleInjectorScopedRequest>();
+                        options.UseMiddleware<JwtTokenMiddleware>();
+                        options.UseMiddleware<ActorMiddleware>();
```
Comment (defectiveAi) [inline source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Program.cs:38]: A health service that requires access to ActorRegistry to determine if access to ActorRegistry is healthy? Verifying that caller has a valid token is fine, but no more than that.

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/FluentHealth.cs (hunk 5 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/FluentHealth.cs]: Fixed

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/Health.cs (hunk 7 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.PostOffice.EntryPoint.Operations/HealthCheck/Health.cs]: fixed

File: source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Program.cs (hunk 8 above)
Comment (michaelmajgaard) [reply source/Energinet.DataHub.PostOffice.EntryPoint.Operations/Program.cs]: fixed

**Review by defectiveAi — APPROVED** [review]
Approved. Stay healthy!

**Pull request merged**